dotcom-rendering/src/web/components/Liveness.importable.tsx

```diff
@@ -0,0 +1,186 @@
+	url.searchParams.set('filterKeyEvents', filterKeyEvents ? 'true' : 'false');
+	return url.href;
+}
+
+const topOfBlog: Element | null | false =
+	!isServer && window.document.querySelector('[data-gu-marker=top-of-blog]');
+
+/**
+ * This allows us to make decisions in javascript based on if the reader
+ * has the top of the blog in view or not
+ *
+ * @returns boolean
+ */
+function topOfBlogVisible(): boolean {
+	return !!topOfBlog && !!topOfBlog.classList.contains('in-viewport');
+}
+
+export const Liveness = ({ pageId, webTitle, ajaxUrl }: Props) => {
+	const [showToast, setShowToast] = useState(false);
+	const [noOfNewPosts, setNoOfNewPosts] = useState(0);
+	const [latestBlockId, setLatestBlockId] = useState(
+		// By default we use the first (latest) block id on the page
+		document.querySelector('#maincontent :first-child')?.id || '',
+	);
+
+	// Read current url to get filterKeyEvents
+	const params = new URLSearchParams(window.location.search);
+	const filterKeyEvents = !!params.get('filterKeyEvents');
+
+	// useApi returns { data, loading, error } but we're not using them here
+	useApi<{
+		numNewBlocks: number;
+		html: string;
+	}>(getKey(pageId, ajaxUrl, filterKeyEvents, latestBlockId), {
+		refreshInterval: 15000,
+		refreshWhenHidden: true,
+		// onSuccess runs (once) after every successful api call. This is useful because it
+		// allows us to avoid the problems of imperative code being executed multiple times
+		// inside react's declarative structure (things get re-rendered when any state changes)
+		onSuccess: (data: {
+			numNewBlocks: number;
+			html: string;
+			mostRecentBlockId: string;
+		}) => {
+			if (data && data.numNewBlocks && data.numNewBlocks > 0) {
+				// Always insert the new blocks in the dom (but hidden)
+				insertNewBlocks(data.html);
+				hydrateBlocks();
+
+				if (topOfBlogVisible()) {
+					revealNewBlocks();
+					setNoOfNewPosts(0);
+				} else {
+					setShowToast(true);
+					// Increment the count of new posts
+					setNoOfNewPosts(noOfNewPosts + data.numNewBlocks);
+				}
+
+				// Update the block id we use for polling
+				if (data.mostRecentBlockId)
+					setLatestBlockId(data.mostRecentBlockId);
+			}
+		},
+	});
+
+	useEffect(() => {
+		document.title =
+			noOfNewPosts > 0 ? `(${noOfNewPosts}) ${webTitle}` : webTitle;
+	}, [noOfNewPosts, webTitle]);
+
+	if (topOfBlog) {
+		const observer = new window.IntersectionObserver(
+			([entry]) => {
+				if (entry.isIntersecting) {
+					entry.target.classList.add('in-viewport');
+					revealNewBlocks();
+					setNoOfNewPosts(0);
+					return;
+				}
+				entry.target.classList.remove('in-viewport');
+			},
+			{
+				root: null,
+				// A margin makes more sense because the element we're
+				// observing (topOfBlog) has no height
+				rootMargin: '100px',
+			},
+		);
+
+		observer.observe(topOfBlog);
+	}
+
+	const handleToastClick = () => {
+		setShowToast(false);
+		if (topOfBlog) topOfBlog.scrollIntoView();
```

Comment:
```suggestion
		if (!topOfBlog) topOfBlog.scrollIntoView();
```

Reply:
Thanks

Reply:
I think because `topOfBlog` is a mix of `Element`, `null` _and_ `false`, that'll narrow it to `null | false` within the `if` statement, which won't allow you to call `scrollIntoView`.

I've dropped a suggestion in above for tweaking the implementation of `topOfBlog`. If you go with that I think you can make this:

```suggestion
		topOfBlog?.scrollIntoView();
```

Alternatively, if you only want to call this if `topOfBlog` is out of the viewport atm:

```suggestion
		if (!topOfBlogVisible()) { topOfBlog?.scrollIntoView(); }
```

Reply:
Implemented locally 👍